qiskit/transpiler/passes/optimization/optimize_1q_decomposition.py

```diff
@@ -92,50 +156,82 @@ def _substitution_checks(self, dag, old_run, new_circ, new_basis):
             return False
 
         # do we even have calibrations?
-        has_cals_p = dag.calibrations is not None and len(dag.calibrations) > 0
+        has_cals = dag.calibrations is not None and len(dag.calibrations) > 0
         # is this run in the target set of this particular decomposer and also uncalibrated?
-        rewriteable_and_in_basis_p = all(
-            g.name in new_basis and (not has_cals_p or not dag.has_calibration_for(g))
+        rewriteable_and_in_basis = all(
+            g.name in new_basis and (not has_cals or not dag.has_calibration_for(g))
             for g in old_run
         )
         # does this run have uncalibrated gates?
-        uncalibrated_p = not has_cals_p or any(not dag.has_calibration_for(g) for g in old_run)
+        uncalibrated = not has_cals or any(not dag.has_calibration_for(g) for g in old_run)
         # does this run have gates not in the image of ._decomposers _and_ uncalibrated?
-        uncalibrated_and_not_basis_p = any(
-            g.name not in self._target_basis and (not has_cals_p or not dag.has_calibration_for(g))
+        uncalibrated_and_not_basis = any(
+            g.name not in self._target_basis and (not has_cals or not dag.has_calibration_for(g))
             for g in old_run
         )
 
-        if rewriteable_and_in_basis_p and len(old_run) < len(new_circ):
-            # NOTE: This is short-circuited on calibrated gates, which we're timid about
-            #       reducing.
-            logger.debug(
+        old_infidelity = 0
+        new_infidelity = 0
+        # look up physical qubit index in layout
+        # (this snippet is adapted from UnitarySynthesis._synth_natural_direction)
+        qubit = next(
+            (index for (index, qubit) in enumerate(dag.qubits) if qubit == old_run[0].qargs[0]),
+            None,
+        )
+
+        if self._backend_properties is not None:
+            pessimistic_fidelities = self._generate_fidelities_dict(qubit, default=-float("inf"))
+            optimistic_fidelities = self._generate_fidelities_dict(qubit)
+
+            for op in old_run:
+                if not isinstance(op, DAGOpNode):
+                    continue
+                old_infidelity += 1.0 - pessimistic_fidelities[op.op.name]
+
+            for instr, _, _ in new_circ.data:
+                new_infidelity += 1.0 - optimistic_fidelities[instr.name]
+
+        # incorporate a possible trace distance from approximate synthesis
+        operator = old_run[0].op.to_matrix()
+        for gate in old_run[1:]:
+            operator = gate.op.to_matrix().dot(operator)
```

Comment:
Don't we already have this unitary previously calculated? Not sure how expensive this recalculation is, might matter for long runs